internal_ws/ykcompile/src/arch/x86_64/mod.rs

```diff
@@ -123,45 +123,121 @@ macro_rules! binop_add_sub {
+                Location::Reg(src_r) => {
+                    // Handle cases where the second operand was stored in an already clobbered
+                    // register. In these cases we load the value back from the stack.
+                    let (r, borrow_rcx) = if src_r == RAX.code() {
+                        let rax_off = i32::try_from(QWORD_REG_SIZE).unwrap();
                         dynasm!(self.asm
-                            ; $op Rb(r)
+                            ; push rcx
+                            ; mov rcx, [rsp + rax_off]
                         );
-                    }
-                    2 => {
+                        (RCX.code(), true)
+                    } else if size > 1 && src_r == RDX.code() {
+                        let rdx_off = i32::try_from(QWORD_REG_SIZE).unwrap() * 2;
                         dynasm!(self.asm
-                            ; $op Rw(r)
+                            ; push rcx
+                            ; mov rcx, [rsp + rdx_off]
                         );
-                    }
-                    4 => {
+                        (RCX.code(), true)
+                    } else {
+                        (src_r, false)
+                    };
+                    // Zero unused extended operand space.
+                    if size > 1 {
                         dynasm!(self.asm
-                            ; $op Rd(r)
+                            ; xor rdx, rdx
                         );
                     }
-                    8 => {
+                    match size {
+                        1 => {
+                            dynasm!(self.asm
+                                ; and ax, 0xff // Zero unused extended operand space.
+                                ; $op Rb(r)
+                            );
+                        }
+                        2 => {
+                            dynasm!(self.asm
+                                ; $op Rw(r)
+                            );
+                        }
+                        4 => {
+                            dynasm!(self.asm
+                                ; $op Rd(r)
+                            );
+                        }
+                        8 => {
+                            dynasm!(self.asm
+                                ; $op Rq(r)
+                            );
+                        }
+                        _ => unreachable!(format!("{}", SIR.ty(&opnd2.ty()))),
+                    }
+                    // If we borrowed RCX, then restore it.
+                    if borrow_rcx {
                         dynasm!(self.asm
-                            ; $op Rq(r)
+                            ; pop rcx
                         );
                     }
-                    _ => unreachable!(format!("{}", SIR.ty(&opnd2.ty()))),
                 },
                 Location::Mem(..) => todo!(),
                 Location::Const { val, .. } => {
+                    dynasm!(self.asm // FIXME not always.
```

Comment:
"not always" is gnomic! What does it mean?

Reply:
That's left in by accident.

Reply:
OK, I shall look forward to the fix commit :)